keystorage/yubihsm2/src/main/java/tech/pegasys/signers/yubihsm/pkcs11/Pkcs11Session.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.signers.yubihsm.pkcs11;
+
+import tech.pegasys.signers.yubihsm.YubiHsmException;
+
+import iaik.pkcs.pkcs11.Module;
+import iaik.pkcs.pkcs11.Session;
+import iaik.pkcs.pkcs11.Slot;
+import iaik.pkcs.pkcs11.Token;
+import iaik.pkcs.pkcs11.TokenException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class Pkcs11Session implements AutoCloseable {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final Session session;
+
+  public Pkcs11Session(final Pkcs11Module module, final Pkcs11YubiHsmPin pin) {
+    final Session session = openReadOnlySession(getToken(module.getModule()));
+
+    try {
+      session.login(Session.UserType.USER, pin.getPin());
+    } catch (final TokenException e) {
+      LOG.error("YubiHSM Login failed {}", e.getMessage());
+      closeSession(session);
+      throw new YubiHsmException("Login Failed", e);
+    }
+    this.session = session;
+  }
+
+  public Session getSession() {
```

Comment:
Shouldn't just open up your session to everyone - removes benefits of encapsulation - can you move the users of this function into here?